src/capture.rs

```diff
@@ -338,26 +352,34 @@ pub mod source {
         let activator = scope.sync_activator_for(&address);
         let activator2 = scope.activator_for(&address);
         let activations = scope.activations();
+        let drop_activator = Arc::new(SyncActivateOnDrop::new((), scope.sync_activator_for(&address)));
         let mut source = source_builder(activator);
         let (mut updates_out, updates) = messages_op.new_output();
         let (mut progress_out, progress) = messages_op.new_output();
-        let tokens_mut = &mut tokens;
+        let token_mut = &mut token;
         messages_op.build(move |capabilities| {
+
+            // Vector of strong references to capabilities, which can be dropped to terminate the sources.
+            let drop_activator_weak = Arc::downgrade(&drop_activator);
+
+            *token_mut = Some(drop_activator);
+
             // Read messages from some source; shuffle them to UPDATES and PROGRESS; share capability with FEEDBACK.
             // First, wrap capabilities in a rc refcell so that they can be downgraded to weak references.
-            use timely::scheduling::activate::ActivateOnDrop;
             let capability_sets = (CapabilitySet::from_elem(capabilities[0].clone()), CapabilitySet::from_elem(capabilities[1].clone()));
             let capability_sets = ActivateOnDrop::new(capability_sets, Rc::new(address), activations);
-            let strong_capabilities = Rc::new(RefCell::new(capability_sets));
-            let local_capabilities = Rc::downgrade(&strong_capabilities);
-            tokens_mut.push(strong_capabilities);
+            let mut local_capabilities = Some(capability_sets);
```

Comment:
I think you can just leave them as sets, and call `downgrade([])` rather than setting the option to `None`.